File: Programmers/Level2/타겟넘버/타겟넘버/main.cpp
//
//  main.cpp
//  타겟넘버
//
//  Created by 홍진석 on 2021/06/23.
//

#include <string>
#include <vector>

using namespace std;

int makeTarget(vector<int> &numbers, int target, int sum, int cnt){
    if(cnt == numbers.size()){
        if(sum == target){
           return 1;
        }
        return 0;
    }
    else{
        int a = makeTarget(numbers, target, sum+numbers[cnt], cnt+1);
        int b = makeTarget(numbers, target, sum-numbers[cnt], cnt+1);
        
        return a+b;
    }
}

int solution(vector<int> numbers, int target) {
    int answer = 0;
    
    answer = makeTarget(numbers, target, 0, 0);
    
    return answer;
}

File: Programmers/Swift/Level2/타겟넘버/타겟넘버/main.swift
//
//  main.swift
//  타겟넘버
//
//  Created by 홍진석 on 2021/09/24.
//

import Foundation

var answer : Int = 0
var numbers : [Int] = [1, 1, 1, 1, 1]
var target = 3
print(solution(numbers, target))

func dfs(_ numbers:[Int], _ target:Int, _ sum:Int, _ cnt:Int) -> Void {
    
    if cnt == numbers.count {
        if sum == target{ answer += 1 }
        return
    }
    
    else{
        dfs(numbers, target, sum+numbers[cnt], cnt+1)
        dfs(numbers, target, sum-numbers[cnt], cnt+1)
    }
}

func solution(_ numbers:[Int], _ target:Int) -> Int {
    dfs(numbers, target, 0, 0)
    return answer
}
